File: LeetCode_Python/Test113.py
__author__ = 'zyt'


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @param {integer} sum
    # @return {integer[][]}
    def pathSum(self, root, sum):
        res, cur = [], []
        if root:
            self.pathSum2(root, sum, cur, res)
        return res

    def pathSum2(self, root, sum, cur, res):
        cur.append(root.val)
        if not root.left and not root.right and root.val == sum:
            res.append(cur[:])
        if root.left:
            self.pathSum2(root.left, sum - root.val, cur, res)
        if root.right:
            self.pathSum2(root.right, sum - root.val, cur, res)
        cur.pop()

File: LeetCode_Java/src/Test113.java
import java.util.ArrayList;
import java.util.List;

/**
 * Created by zyt on 15/8/19 20:01.
 */
public class Test113 {

    /**
     * Definition for a binary tree node.
     */
    private class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public class Solution {
        public List<List<Integer>> pathSum(TreeNode root, int sum) {
            List<List<Integer>> result = new ArrayList<>();
            List<Integer> current = new ArrayList<>();
            if (root == null) {
                return result;
            }
            return this.pathSum(root, sum, current, result);
        }

        public List<List<Integer>> pathSum(TreeNode root, int sum, List<Integer> current, List<List<Integer>> result) {
            current.add(root.val);
            if (root.left == null && root.right == null && root.val == sum) {
                result.add(new ArrayList<>(current));
            }
            if (root.left != null) {
                this.pathSum(root.left, sum - root.val, current, result);
            }
            if (root.right != null) {
                this.pathSum(root.right, sum - root.val, current, result);
            }
            current.remove(current.size() - 1);
            return result;
        }
    }
}
